lib/awful/completion.lua

```diff
@@ -154,10 +161,25 @@ function completion.shell(command, cur_pos, ncomp, shell)
         while true do
             local line = c:read("*line")
             if not line then break end
-            if os.execute("test -d " .. string.format('%q', line)) == 0 then
-                line = line .. "/"
+            if comptype == "command" then
+                if str_starts(line, ".") or str_starts(line, "/") then
+                    if gfs.file_executable(line) then
+                        table.insert(output, bash_escape(line))
+                    end
+                else
+                    for path in string.gmatch(os.getenv('PATH'), "[^:]+") do
+                        if gfs.file_executable(path .. "/" .. line) then
+                            table.insert(output, bash_escape(line))
+                            break
+                        end
+                    end
+                end
+            end
+            if gfs.is_dir(line) then
+                table.insert(output, bash_escape(line .. "/"))
+            elseif comptype == "file" then
+                table.insert(output, bash_escape(line))
             end
-            table.insert(output, bash_escape(line))
```

Comment:
> Wouldn't commands (and aliases for this sake) be enough? Is it really necessary to add directories to the completion results?

Making it simpler would be good IMHO.
Directories should get completed after `./` then.

Current master does not complete (local) directories for me in Zsh, i.e. `Down` does not complete "Downloads", but `./Down` does.
This is what I would expect for bash then, too.